File: Regular-Mocks /DSA/CustomSortString.java
// PS: https://leetcode.com/problems/custom-sort-string/

import java.util.Map;
import java.util.HashMap;

class Solution {
    /*
        1. HashMap for string s
        2. Move in order and fill up resultant if same char in s
        3. Fill in rest of chars
    */
    public String customSortString(String order, String s) {
        Map<Character, Integer> map = new HashMap();
        StringBuilder result = new StringBuilder();
        
        for(char c: s.toCharArray()){
            if (map.containsKey(c)){
                map.put(c, map.get(c) + 1);
            }else{
                map.put(c, 1);
            }
        }

        for(char c : order.toCharArray()){
            if (map.containsKey(c)){
                while (map.get(c) != 0){
                    result.append(c);
                    map.put(c, map.get(c) - 1);    
                }
            }
        }

        // Now add all those elements which are not present in order.
        for(char key : map.keySet()){
            while (map.get(key) != 0){
                result.append(key);
                map.put(key, map.get(key) - 1);
            }
        }


        return result.toString();
    
    }
}

File: Regular-Mocks /DSA/CustomSortString.py
# PS: https://leetcode.com/problems/custom-sort-string
class Solution:
    """
    
        T.C: O(N), S.C: (N)
        
        1. Map out all the chars and it's frequency from String s and store it in a map
        2. Travserse Order string and check if that character is present in String S if yes add that to resultant string
        3. Now add rest of the characters from String s which are not present in String order
    
    """
    def customSortString(self, order: str, s: str) -> str:
        mapS = {}
        
        for char in s:
            mapS[char] = mapS.get(char, 0) + 1

        # make sure the order remains the same
        result = []

        for key in order:
            if key in mapS:
                while mapS[key]:
                    result.append(key)
                    mapS[key] -= 1
        
        for k, v in mapS.items():
            while mapS[k]:
                result.append(k)
                mapS[k] -= 1


        return ''.join(result)
         
